lib/manager/components/transform/AddCustomFile.js

```diff
@@ -0,0 +1,98 @@
+// @flow
+
+import React, { Component } from 'react'
+
+import type { AddCustomFileProps, TransformProps } from '../../../types'
+import CSV_VALIDATION_ERRORS from '../../util/enums/transform'
+import { getComponentMessages } from '../../../common/util/config'
+
+import CustomCSVForm from './CustomCSVForm'
+
+/**
+ * Component that renders fields for AddCustomFile. This transformation shares csvData props with the ReplaceFileFromString transformation.
+ * TODO: adapt this transformation to include a file upload for larger custom files?
+ */
+export default class AddCustomFile extends Component<TransformProps<AddCustomFileProps>, AddCustomFileProps> {
+  // Messages are for the child CSV Form component but since messages are shared across transformation types,
+  // the messages are being grouped under that component.
+  messages = getComponentMessages('AddCustomFile')
+  constructor (props: TransformProps<AddCustomFileProps>) {
+    super(props)
+    this.state = {csvData: props.transformation.csvData, table: props.transformation.table}
+  }
+
+  componentDidMount () {
+    this._updateErrors()
+  }
+
+  componentDidUpdate (prevProps: TransformProps<AddCustomFileProps>, prevState: AddCustomFileProps) {
+    if (prevState !== this.state) {
+      this._updateErrors()
+    }
+  }
+
+  _handleChange = (evt: SyntheticInputEvent<HTMLInputElement>) => {
+    const newState = {...this.state, table: evt.target.value}
+    this.setState(newState)
+  }
+
+  _onSaveCsvData = () => {
+    const csvData = this.state.csvData || null
+    const table = this.state.table
+    this.props.onSave({csvData, table}, this.props.index)
+  }
+
+  _onChangeCsvData = (evt: SyntheticInputEvent<HTMLInputElement>) => {
+    const newState = {...this.state, csvData: evt.target.value}
+    this.setState(newState)
+  }
+
+  _getValidationErrors (fields: AddCustomFileProps): Array<string> {
+    const issues = []
+    const { csvData, table } = fields
+
+    // CSV data must be defined.
+    if (!csvData || csvData.length === 0) {
+      issues.push(CSV_VALIDATION_ERRORS.UNDEFINED_CSV_DATA)
+    }
+    if (!table) {
```

Comment:
There should also be a check to ensure that the custom file name does not contain ".txt".

Reply:
Good call, added that in https://github.com/ibi-group/datatools-ui/pull/950/commits/2d62c9f0abb36da9923c66e1543ed615abd20ffa